### DISCUSSION TITLE
How does a GraphQL query/mutation string get converted into a DataFetchingEnvironment

### DISCUSSION THREAD

**Discussion by spierepf** [graphql-java/graphql-java#2720] [Q&A]
I am writing a Grails 5 app using a GraphQL plugin. Part of that app needs to place restrictions on the queries and mutations it allows to proceed.

The Grails plugin offers a `GraphQLFetcherInterceptor` class that implementors can use to decide which queries get processed. The interceptor receives a DataFetchingEnvironment and is expected to return a boolean to indicate whether the query should proceed.

What I would like to do is write an array of integration tests that accept a graphql query/mutation string, parse it into a DataFetchingEnvironment, then pass that to the interceptor and compare the return value with an expected true or false.

The only step I'm missing is how a query string is parsed into a `DataFetchingEnvironment`.

Can anyone elucidate?

Thanks!

**Comment by bbakerman** [comment]
> Part of that app needs to place restrictions on the queries and mutations it allows to proceed.

It's probably too late to wait for the `DataFetchingEnvironment` = a DFE is created when the engine executes each field.  Its well after the query level

I would suggest `graphql.execution.instrumentation.Instrumentation` and specifically the `graphql.execution.instrumentation.Instrumentation#beginExecuteOperation`

This is called when the query is parsed and validated.

https://www.graphql-java.com/documentation/instrumentation

if you have a set of allowed ONLY queries - then there is also `graphql.execution.preparsed.PreparsedDocumentProvider` which is a caching mechanism that can also allow only certain queries (and error out if not the ones you like)

https://www.graphql-java.com/documentation/execution/#query-caching
